library/std/src/f32.rs

```diff
@@ -876,4 +876,32 @@ impl f32 {
     pub fn atanh(self) -> f32 {
         0.5 * ((2.0 * self) / (1.0 - self)).ln_1p()
     }
+
+    /// Linear interpolation between `start` and `end`.
+    ///
+    /// This enables the calculation of a "smooth" transition between `start` and `end`,
+    /// where start is represented by `self == 0.0` and `end` is represented by `self == 1.0`.
+    ///
+    /// Values below 0.0 or above 1.0 are allowed, and in general this function closely
+    /// resembles the value of `start + self * (end - start)`, plus additional guarantees.
+    ///
+    /// Those guarantees are, assuming that all values are [`finite`]:
```

Comment:
Can we give any guarantees about non-finite values? E.g. that any nan wil result in a nan result, and/or that if only one of the values is infinite, the result will be that infinite value.

Reply:
So, I was considering doing the full list of guarantees that [C++ provides](https://en.cppreference.com/w/cpp/numeric/lerp), which are:

* If isfinite(a) && isfinite(b):
    * If t == 0, the result is equal to a.
    * If t == 1, the result is equal to b.
    * If t >= 0 && t <= 1, the result is finite.
* If isfinite(t) && a == b, the result is equal to a.
* If isfinite(t) || (!isnan(t) && b-a != 0), the result is not NaN.
* Let CMP(x,y) be 1 if x > y, -1 if x < y, and 0 otherwise. For any t1 and t2, the product of CMP(lerp(a, b, t2), lerp(a, b, t1)), CMP(t2, t1), and CMP(b, a) is non-negative. (That is, lerp is monotonic.)

But I figured that it might potentially limit what kinds of implementations we allow. I'll try adding tests for these (and comment them out if they don't work) and maybe we can hash it out in the tracking issue.